eruditus/msg_components/forms/credentials.py

```diff
@@ -0,0 +1,170 @@
+from urllib.parse import urlparse, parse_qs
+
+import discord
+from config import MONGO, DBNAME, CTF_COLLECTION
+
+
+class CTFdCredentialsForm(discord.ui.Modal, title="Add CTFd credentials"):
+    username = discord.ui.TextInput(
+        label="Username",
+        style=discord.TextStyle.short,
+        placeholder="Enter your username...",
+        required=True,
+        max_length=128,
+    )
+    password = discord.ui.TextInput(
+        label="Password",
+        style=discord.TextStyle.short,
+        placeholder="Enter your password...",
+        required=True,
+        max_length=128,
+    )
+
+    def __init__(self, url: str) -> None:
+        super().__init__()
+        self.url = url
+
+    async def on_submit(self, interaction: discord.Interaction) -> None:
+        ctf = MONGO[DBNAME][CTF_COLLECTION].find_one(
+            {"guild_category": interaction.channel.category_id}
+        )
+        ctf["credentials"]["url"] = self.url
+        ctf["credentials"]["username"] = self.username.value
+        ctf["credentials"]["password"] = self.password.value
+
+        MONGO[DBNAME][CTF_COLLECTION].update_one(
+            {"_id": ctf["_id"]},
+            {"$set": {"credentials": ctf["credentials"]}},
+        )
+
+        creds_channel = discord.utils.get(
+            interaction.guild.text_channels, id=ctf["guild_channels"]["credentials"]
+        )
+        message = (
+            f"CTFd platform: {self.url}\n"
+            "```yaml\n"
+            f"Username: {self.username.value}\n"
+            f"Password: {self.password.value}\n"
+            "```"
+        )
+
+        await creds_channel.purge()
+        await creds_channel.send(message, suppress_embeds=True)
+        await interaction.response.send_message("✅ Credentials added.", ephemeral=True)
+
+
+class RCTFCredentialsForm(discord.ui.Modal, title="Add rCTF invite link"):
+    invite = discord.ui.TextInput(
+        label="Invite link",
+        style=discord.TextStyle.short,
+        placeholder="https://rctf.example.com/login?token=<token>",
+        required=True,
+        max_length=1024,
+    )
+
+    def __init__(self, url: str) -> None:
+        super().__init__()
+        self.url = url
+
+    async def on_submit(self, interaction: discord.Interaction) -> None:
+        parsed_url = urlparse(self.invite.value)
+        params = parse_qs(parsed_url.query)
+        if not (token := params.get("token")):
+            await interaction.response.send_message(
+                "Token was not found in the URL, please submit a valid invite link.",
+                ephemeral=True,
+            )
+            return
+        token = token[0]
+
+        ctf = MONGO[DBNAME][CTF_COLLECTION].find_one(
+            {"guild_category": interaction.channel.category_id}
+        )
+        ctf["credentials"]["url"] = self.url
+        ctf["credentials"]["invite"] = self.invite.value
```

Comment:
If i understand correctly it would save the invite in format `https://domain.com/invite?token=AAAAA`, but i think the rCTF platform module requires it to be just `AAAA`. So either we should parse the token value from url or? Looks weird but maybe its just me.

Also, the `token` value here should be a bearer token value IIRC.

Reply:
We will have both in the credentials (`invite`: the entire URL, and `token` which is parsed from the URL in the above code, we will change its name to `teamToken`). Then `authToken` can only be set in the session context once we login for the first time.
What do you think?

Reply:
Sounds good to me.